core/src/main/java/org/elasticsearch/index/translog/TranslogService.java

```diff
@@ -118,37 +124,64 @@ private TimeValue computeNextInterval() {
         return new TimeValue(interval.millis() + (ThreadLocalRandom.current().nextLong(interval.millis())));
     }
 
-    private class TranslogBasedFlush implements Runnable {
+    // public for testing
+    public TranslogBasedFlush createTranslogBasedFlush() {
+        return new TranslogBasedFlush();
+    }
+
+    public class TranslogBasedFlush extends AbstractRunnable {
 
         private volatile long lastFlushTime = System.currentTimeMillis();
 
```

Comment:
we could add a `volatile boolean` here to indicate that we are either running or are scheduled and assert on that when we close the shard? just and idea to ensure we don't miss anything along the lines?